packages/beacon-node/src/api/impl/validator/index.ts

```diff
@@ -490,196 +513,185 @@ export function getValidatorApi({
-              timeoutMs: BLOCK_PRODUCTION_RACE_TIMEOUT_MS,
-              slot,
-            });
-          }
-        );
-        if (blindedBlock instanceof Error) {
-          // error here means race cutoff exceeded
-          logger.error("Failed to produce builder block", {slot}, blindedBlock);
-          blindedBlock = null;
-        }
-        if (fullBlock instanceof Error) {
-          logger.error("Failed to produce execution block", {slot}, fullBlock);
-          fullBlock = null;
+      const builderDisabledError = new Error("Builder disabled");
+      const engineDisabledError = new Error("Engine disabled");
+      // use abort controller to stop waiting for both block sources
+      const controller = new AbortController();
+      const [builder, engine] = await resolveOrRacePromises(
+        [
+          isBuilderEnabled
+            ? // can't do fee recipient checks as builder bid doesn't return feeRecipient as of now
+              produceBuilderBlindedBlock(slot, randaoReveal, graffiti, {
+                feeRecipient,
+                // skip checking and recomputing head in these individual produce calls
+                skipHeadChecksAndUpdate: true,
+                commonBlockBody,
+              })
+            : Promise.reject<ReturnType<typeof produceBuilderBlindedBlock>>(builderDisabledError),
+
+          isEngineEnabled
+            ? (produceEngineFullBlockOrContents(slot, randaoReveal, graffiti, {
+                feeRecipient,
+                strictFeeRecipientCheck,
+                // skip checking and recomputing head in these individual produce calls
+                skipHeadChecksAndUpdate: true,
+                commonBlockBody,
+              }).then((engine) => {
+                // Once the engine returns a block, in the event of either:
+                // - suspected builder censorship
+                // - sufficiently small builder boost
+                // we should not wait for the builder promise to resolve
+                if (engine.shouldOverrideBuilder || builderBoostFactor === BigInt(0)) {
+                  controller.abort();
+                }
+                return engine;
+              }) as ReturnType<typeof produceEngineFullBlockOrContents>)
+            : Promise.reject<ReturnType<typeof produceEngineFullBlockOrContents>>(engineDisabledError),
+        ],
+        {
+          resolveTimeoutMs: BLOCK_PRODUCTION_RACE_CUTOFF_MS,
+          raceTimeoutMs: BLOCK_PRODUCTION_RACE_TIMEOUT_MS,
+          signal: controller.signal,
         }
-      } else if (blindedBlockPromise !== null && fullBlockPromise === null) {
-        blindedBlock = await blindedBlockPromise;
-        fullBlock = null;
-      } else if (blindedBlockPromise === null && fullBlockPromise !== null) {
-        blindedBlock = null;
-        fullBlock = await fullBlockPromise;
-      } else {
+      );
+
+      if (
+        builder.status === "rejected" &&
+        builder.reason === builderDisabledError &&
```

Comment:
This reads a bit wrong to me, builder disabled shouldn't be considered an error even though technically we use an error to determine that it was disabled

Simple renaming would make this clearer

```suggestion
        builder.reason === builderDisabled &&
```